Error inside networkx

Steps to reproduce:

```
firedrake-install
. firedrake/bin/activate
git clone git@github.com:thetisproject/thetis
pip install -e thetis
pip install colorama
pip install scipy
cd thetis/test/swe2d
firedrake-clean
py.test test_steady_state_channel_mms.py -x -v -s
============================= test session starts ==============================
platform darwin -- Python 2.7.11, pytest-2.9.1, py-1.4.31, pluggy-0.3.1 -- /Users/lmitche1/Documents/work/src/fd-install/firedrake/bin/python2.7
cachedir: ../../.cache
benchmark: 3.0.0 (defaults: timer=time.time disable_gc=False min_rounds=5 min_time=5.00us max_time=1.00s calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /Users/lmitche1/Documents/work/src/fd-install/thetis, inifile: 
plugins: benchmark-3.0.0, xdist-1.14
collecting 0 items--------------------------------------------------------------------------
Petsc Development GIT revision: v3.4.2-12173-g511e85b  GIT Date: 2016-04-07 00:39:56 -0500
       The PETSc Team
    petsc-maint@mcs.anl.gov
 http://www.mcs.anl.gov/petsc/
See docs/changes/index.html for recent updates.
See docs/faq.html for problems.
See docs/manualpages/index.html for help. 
Libraries linked from /Users/lmitche1/Documents/work/src/fd-install/firedrake/lib/python2.7/site-packages/petsc/lib
--------------------------------------------------------------------------
collected 2 items 

test_steady_state_channel_mms.py::test_steady_state_channel_mms[mimetic] COFFEE finished in 0.00221896 seconds (flops: 0 -> 0)
COFFEE finished in 0.00326109 seconds (flops: 0 -> 0)
compute_form_data finished in 0.024508 seconds.
compile_integral finished in 0.0944471 seconds.
TSFC finished in 0.119004 seconds.
COFFEE finished in 0.0517299 seconds (flops: 4340 -> 1316)
COFFEE finished in 0.00372005 seconds (flops: 0 -> 0)
compute_form_data finished in 0.0209169 seconds.
compile_integral finished in 0.1691 seconds.
TSFC finished in 0.190075 seconds.
COFFEE finished in 0.0528538 seconds (flops: 1757 -> 749)
COFFEE finished in 0.00235796 seconds (flops: 0 -> 0)
COFFEE finished in 0.00304103 seconds (flops: 0 -> 0)
compute_form_data finished in 0.0181992 seconds.
compile_integral finished in 0.0292919 seconds.
TSFC finished in 0.047559 seconds.
COFFEE finished in 0.00383091 seconds (flops: 27 -> 27)
compute_form_data finished in 0.0216379 seconds.
compile_integral finished in 0.0289121 seconds.
TSFC finished in 0.0506079 seconds.
COFFEE finished in 0.0046041 seconds (flops: 27 -> 27)
compute_form_data finished in 0.019644 seconds.
compile_integral finished in 0.028259 seconds.
TSFC finished in 0.047955 seconds.
COFFEE finished in 0.0049181 seconds (flops: 27 -> 27)
compute_form_data finished in 0.022316 seconds.
compile_integral finished in 0.0296991 seconds.
TSFC finished in 0.052078 seconds.
COFFEE finished in 0.00385499 seconds (flops: 27 -> 27)
COFFEE finished in 0.00276399 seconds (flops: 0 -> 0)
COFFEE finished in 0.00211 seconds (flops: 0 -> 0)
dt = 10.0
compute_form_data finished in 0.872862 seconds.
compile_integral finished in 0.323892 seconds.
compile_integral finished in 0.125879 seconds.
compile_integral finished in 0.120333 seconds.
compile_integral finished in 0.112606 seconds.
compile_integral finished in 0.112715 seconds.
compile_integral finished in 0.303551 seconds.
TSFC finished in 1.97202 seconds.
COFFEE finished in 0.312116 seconds (flops: 21956 -> 10336)
COFFEE finished in 0.270186 seconds (flops: 18129 -> 7332)
COFFEE finished in 0.314766 seconds (flops: 18899 -> 12557)
COFFEE finished in 0.291351 seconds (flops: 6137 -> 4354)
COFFEE finished in 0.303944 seconds (flops: 6137 -> 4354)
FAILED

=================================== FAILURES ===================================
____________________ test_steady_state_channel_mms[mimetic] ____________________

options = {'no_exports': True}

    @pytest.mark.parametrize("options", [
        {"no_exports": True},  # default: mimetic
        {"no_exports": True, "mimetic": False, "continuous_pressure": True},
    ], ids=["mimetic", "pndp(n+1)"])
    def test_steady_state_channel_mms(options):
        lx = 5e3
        ly = 1e3
    
        order = 1
        # minimum resolution
        min_cells = 16
        n = 1  # number of timesteps
        dt = 10.
        g = physical_constants['g_grav'].dat.data[0]
        h0 = 10.  # depth at rest
        area = lx*ly
    
        k = 4.0*math.pi/lx
        q = h0*1.0  # flux (depth-integrated velocity)
        eta0 = 1.0  # free surface amplitude
    
        eta_expr = Expression("eta0*cos(k*x[0])", k=k, eta0=eta0)
        depth_expr = "H0+eta0*cos(k*x[0])"
        u_expr = Expression(("Q/({H})".format(H=depth_expr), 0.), k=k, Q=q, eta0=eta0, H0=h0)
        # source_expr = Expression("k*eta0*(pow(Q,2)/pow({H},3)-g)*sin(k*x[0])".format(H=depth_expr),
        #                          k=k, g=g, Q=q, eta0=eta0, H0=h0)
        source_expr = Expression(("k*eta0*(pow(Q,2)/pow({H},3)-g)*sin(k*x[0])".format(H=depth_expr), 0),
                                 k=k, g=g, Q=q, eta0=eta0, H0=h0)
        u_bcval = q/(h0+eta0)
        eta_bcval = eta0
    
        do_exports = not options['no_exports']
        if do_exports:
            diff_pvd = File('diff.pvd')
            udiff_pvd = File('udiff.pvd')
            source_pvd = File('source.pvd')
    
        eta_errs = []
        u_errs = []
        for i in range(5):
            mesh2d = RectangleMesh(min_cells*2**i, 1, lx, ly)
    
            # bathymetry
            p1_2d = FunctionSpace(mesh2d, 'CG', 1)
            bathymetry_2d = Function(p1_2d, name="bathymetry")
            bathymetry_2d.assign(h0)
    
            source_space = VectorFunctionSpace(mesh2d, 'DG', order+1)
            source_func = project(source_expr, source_space)
    
            # --- create solver ---
            solver_obj = solver2d.FlowSolver2d(mesh2d, bathymetry_2d, order=order)
            solver_obj.options.nonlin = True
            solver_obj.options.t_export = dt
            solver_obj.options.t_end = n*dt
            solver_obj.options.uv_source_2d = source_func
            solver_obj.options.timestepper_type = 'cranknicolson'
            solver_obj.options.timer_labels = []
            solver_obj.options.dt = dt
            solver_obj.options.update(options)
    
            # boundary conditions
            inflow_tag = 1
            outflow_tag = 2
    
            inflow_func = Function(p1_2d)
            inflow_func.interpolate(Expression(-u_bcval))
            inflow_bc = {'un': inflow_func}
            outflow_func = Function(p1_2d)
            outflow_func.interpolate(Expression(eta_bcval))
            outflow_bc = {'elev': outflow_func}
            solver_obj.bnd_functions['shallow_water'] = {inflow_tag: inflow_bc, outflow_tag: outflow_bc}
            # parameters['quadrature_degree']=5
    
>           solver_obj.create_equations()

test_steady_state_channel_mms.py:81: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../thetis/solver2d.py:143: in create_equations
    semi_implicit=self.options.use_linearized_semi_implicit_2d)
../../thetis/timeintegrator.py:503: in __init__
    self.update_solver()
../../thetis/timeintegrator.py:510: in update_solver
    options_prefix=self.name)
../../../firedrake/src/firedrake/firedrake/variational_solver.py:125: in __init__
    ctx = solving_utils._SNESContext(problem)
../../../firedrake/src/firedrake/firedrake/solving_utils.py:106: in __init__
    for problem in problems)
../../../firedrake/src/firedrake/firedrake/solving_utils.py:106: in <genexpr>
    for problem in problems)
../../../firedrake/src/firedrake/firedrake/assemble.py:66: in assemble
    inverse=inverse, nest=nest)
<decorator-gen-295>:2: in _assemble
    ???
../../../firedrake/src/firedrake/firedrake/utils.py:62: in wrapper
    return f(*args, **kwargs)
../../../firedrake/src/firedrake/firedrake/assemble.py:100: in _assemble
    inverse=inverse)
../../../firedrake/src/firedrake/firedrake/tsfc_interface.py:243: in compile_form
    number_map).kernels
../../../firedrake/src/PyOP2/pyop2/caching.py:203: in __new__
    obj = make_obj()
../../../firedrake/src/PyOP2/pyop2/caching.py:193: in make_obj
    obj.__init__(*args, **kwargs)
../../../firedrake/src/firedrake/firedrake/tsfc_interface.py:174: in __init__
    kernels.append(KernelInfo(kernel=Kernel(ast, ast.name, opts=opts),
../../../firedrake/src/PyOP2/pyop2/backends.py:118: in __call__
    return t(*args, **kwargs)
../../../firedrake/src/PyOP2/pyop2/caching.py:203: in __new__
    obj = make_obj()
../../../firedrake/src/PyOP2/pyop2/caching.py:193: in make_obj
    obj.__init__(*args, **kwargs)
../../../firedrake/src/PyOP2/pyop2/base.py:3848: in __init__
    self._code = self._ast_to_c(self._ast, self._opts)
../../../firedrake/src/PyOP2/pyop2/host.py:60: in _ast_to_c
    ast_handler.plan_cpu(self._opts)
../../../firedrake/src/COFFEE/coffee/plan.py:184: in plan_cpu
    _generate_cpu_code(self, kernel, **params)
../../../firedrake/src/COFFEE/coffee/plan.py:106: in _generate_cpu_code
    loop_opt.rewrite(rewrite)
../../../firedrake/src/COFFEE/coffee/optimizer.py:116: in rewrite
    ew.SGrewrite()
../../../firedrake/src/COFFEE/coffee/rewriter.py:454: in SGrewrite
    sgraph, mapper = sg_visitor.visit(self.stmt.rvalue)
../../../firedrake/src/COFFEE/coffee/visitor.py:106: in visit
    return meth(o, *args, **kwargs)
../../../firedrake/src/COFFEE/coffee/visitors/utilities.py:455: in visit_Sum
    ret = self.visit(op, ret=ret, syms=loc_syms[i], parent=o)
../../../firedrake/src/COFFEE/coffee/visitor.py:106: in visit
    return meth(o, *args, **kwargs)
../../../firedrake/src/COFFEE/coffee/visitors/utilities.py:439: in visit_Prod
    G.add_edges_from(loc_syms)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <networkx.classes.graph.Graph object at 0x1133d1bd0>
ebunch = [(('t1', ('facet[0]', 'ip', 'j'), ((1, 0), (1, 0), (1, 0))), ('t1', ('facet[1]', 'ip', 'k'), ((1, 0), (1, 0), (1, 0)))...]', 'ip', 'j'), ((1, 0), (1, 0), (1, 0))), ('t37', (), ()), ('t0', ('facet[1]', 'ip', 'k'), ((1, 0), (1, 0), (1, 0))))]
attr_dict = {}, attr = {}
e = (('t1', ('facet[0]', 'ip', 'j'), ((1, 0), (1, 0), (1, 0))), ('t37', (), ()), ('t1', ('facet[1]', 'ip', 'k'), ((1, 0), (1, 0), (1, 0))))
ne = 3, u = ('t1', ('facet[0]', 'ip', 'j'), ((1, 0), (1, 0), (1, 0)))
v = ('t37', (), ())
dd = ('t1', ('facet[1]', 'ip', 'k'), ((1, 0), (1, 0), (1, 0)))
datadict = {'t': '1'}

    def add_edges_from(self, ebunch, attr_dict=None, **attr):
        """Add all the edges in ebunch.
    
            Parameters
            ----------
            ebunch : container of edges
                Each edge given in the container will be added to the
                graph. The edges must be given as as 2-tuples (u,v) or
                3-tuples (u,v,d) where d is a dictionary containing edge
                data.
            attr_dict : dictionary, optional (default= no attributes)
                Dictionary of edge attributes.  Key/value pairs will
                update existing data associated with each edge.
            attr : keyword arguments, optional
                Edge data (or labels or objects) can be assigned using
                keyword arguments.
    
    
            See Also
            --------
            add_edge : add a single edge
            add_weighted_edges_from : convenient way to add weighted edges
    
            Notes
            -----
            Adding the same edge twice has no effect but any edge data
            will be updated when each duplicate edge is added.
    
            Edge attributes specified in edges take precedence
            over attributes specified generally.
    
            Examples
            --------
            >>> G = nx.Graph()   # or DiGraph, MultiGraph, MultiDiGraph, etc
            >>> G.add_edges_from([(0,1),(1,2)]) # using a list of edge tuples
            >>> e = zip(range(0,3),range(1,4))
            >>> G.add_edges_from(e) # Add the path graph 0-1-2-3
    
            Associate data to edges
    
            >>> G.add_edges_from([(1,2),(2,3)], weight=3)
            >>> G.add_edges_from([(3,4),(1,4)], label='WN2898')
            """
        # set up attribute dict
        if attr_dict is None:
            attr_dict = attr
        else:
            try:
                attr_dict.update(attr)
            except AttributeError:
                raise NetworkXError(
                    "The attr_dict argument must be a dictionary.")
        # process ebunch
        for e in ebunch:
            ne = len(e)
            if ne == 3:
                u, v, dd = e
            elif ne == 2:
                u, v = e
                dd = {}  # doesnt need edge_attr_dict_factory
            else:
                raise NetworkXError(
                    "Edge tuple %s must be a 2-tuple or 3-tuple." % (e,))
            if u not in self.node:
                self.adj[u] = self.adjlist_dict_factory()
                self.node[u] = {}
            if v not in self.node:
                self.adj[v] = self.adjlist_dict_factory()
                self.node[v] = {}
            datadict = self.adj[u].get(v, self.edge_attr_dict_factory())
            datadict.update(attr_dict)
>           datadict.update(dd)
E           ValueError: dictionary update sequence element #1 has length 3; 2 is required

../../../firedrake/lib/python2.7/site-packages/networkx/classes/graph.py:874: ValueError
```

--- a/coffee/base.py
+++ b/coffee/base.py
@@ -434,7 +434,13 @@ class Symbol(Expr):
 
     :param tuple rank: entries represent the iteration variables the symbol
         depends on, or explicit numbers representing the entry of a tensor the
-        symbol is accessing, or the size of the tensor itself. """
+        symbol is accessing, or the size of the tensor itself.
+    :param list offset: an iterator of 2-tuple (period, stride), one 2-tuple for
+        each entry in rank. The period is the multiplier of the rank, while
+        stride is a quantity summed to the rank. E.g., if rank=(i, j) and
+        offset=((1, 0), (3, 2)), printing a symbol 'a' returns the string
+        a[i][3*j + 2].
+    """
 
     def __init__(self, symbol, rank=(), offset=()):
         super(Symbol, self).__init__([])

--- a/coffee/cse.py
+++ b/coffee/cse.py
@@ -139,13 +139,20 @@ class CSEUnpicker(object):
     some symbols (further information concerning loop linearity is available in
     ``expression.py``)."""
 
-    def __init__(self, stmt, expr_info, header, hoisted, decls, expr_graph):
-        self.stmt = stmt
-        self.expr_info = expr_info
+    def __init__(self, exprs, header, hoisted, decls, expr_graph):
+        self.exprs = exprs
         self.header = header
         self.hoisted = hoisted
         self.decls = decls
         self.expr_graph = expr_graph
+
+    @property
+    def type(self):
+        return self.exprs.values()[0].type
+
+    @property
+    def linear_dims(self):
+        return self.exprs.values()[0].linear_dims
 
     def _push_temporaries(self, temporaries, loop, trace, global_trace):
 
@@ -205,15 +212,15 @@ class CSEUnpicker(object):
         from rewriter import ExpressionRewriter
 
         # Never attempt to transform the main expression
-        temporaries = [t for t in temporaries if t.node != self.stmt]
+        temporaries = [t for t in temporaries if t.node not in self.exprs]
 
         lda = ldanalysis(self.header, key='symbol', value='dim')
 
         # Expand + Factorize
         rewriters = OrderedDict()
         for t in temporaries:
-            expr_info = MetaExpr(self.expr_info.type, t.main_loop.children[0],
-                                 t.nest, tuple(l.dim for l in t.loops if l.is_linear))
+            expr_info = MetaExpr(self.type, t.main_loop.children[0], t.nest,
+                                 tuple(l.dim for l in t.loops if l.is_linear))
             ew = ExpressionRewriter(t.node, expr_info, self.decls, self.header,
                                     self.hoisted, self.expr_graph)
             ew.replacediv()
@@ -231,8 +238,7 @@ class CSEUnpicker(object):
     def _analyze_expr(self, expr, lda):
         finder = FindInstances(Symbol)
         syms = finder.visit(expr, ret=FindInstances.default_retval())[Symbol]
-        syms = [s for s in syms
-                if any(l in self.expr_info.linear_dims for l in lda[s])]
+        syms = [s for s in syms if any(l in self.linear_dims for l in lda[s])]
 
         syms_costs = defaultdict(int)
 

--- a/coffee/cse.py
+++ b/coffee/cse.py
@@ -55,7 +55,7 @@ class Temporary(object):
         self.node = node
         self.main_loop = main_loop
         self.nest = nest
-        self.reads_costs = reads_costs or {}
+        self.reads_costs = reads_costs or OrderedDict()
         self.flops = EstimateFlops().visit(node)
 
     @property
@@ -117,7 +117,7 @@ class Temporary(object):
 
     def reconstruct(self):
         temporary = Temporary(self.node, self.main_loop,
-                              self.nest, dict(self.reads_costs))
+                              self.nest, OrderedDict(self.reads_costs))
         temporary.level = self.level
         temporary.is_read = list(self.is_read)
         return temporary
@@ -240,11 +240,12 @@ class CSEUnpicker(object):
         syms = finder.visit(expr, ret=FindInstances.default_retval())[Symbol]
         syms = [s for s in syms if any(l in self.linear_dims for l in lda[s])]
 
-        syms_costs = defaultdict(int)
+        syms_costs = OrderedDict()
 
         def wrapper(node, found=0):
             if isinstance(node, Symbol):
                 if node in syms:
+                    syms_costs.setdefault(node, 0)
                     syms_costs[node] += found
                 return
             elif isinstance(node, (EmptyStatement, ArrayInit)):

--- a/coffee/cse.py
+++ b/coffee/cse.py
@@ -169,7 +169,8 @@ class CSEUnpicker(object):
             if temporary.is_static_init:
                 # ... its rvalue must not be an array initializer
                 return False
-            if not all(r.pushed or
+            if not all(not r.reads or
+                       r.pushed or
                        loop == r.main_loop or
                        temporary.main_loop.dim in r.rank for r in reads):
                 # ... all the read temporaries must still be accessible

--- a/coffee/hoister.py
+++ b/coffee/hoister.py
@@ -376,9 +376,14 @@ class Hoister(object):
                 # 3) Create the required new AST nodes
                 symbols, decls, stmts = [], [], []
                 for i, e in enumerate(subexprs):
+                    already_hoisted = False
                     if global_cse and self.hoisted.get_symbol(e):
                         name = self.hoisted.get_symbol(e)
-                    else:
+                        decl = self.hoisted[name].decl
+                        if decl in place.children and \
+                                place.children.index(decl) < place.children.index(offset):
+                            already_hoisted = True
+                    if not already_hoisted:
                         name = self._hoisted_sym % {
                             'loop_dep': '_'.join(dep) if dep else 'c',
                             'expr_id': self.expr_id,

--- a/coffee/optimizer.py
+++ b/coffee/optimizer.py
@@ -43,6 +43,7 @@ from base import *
 from utils import *
 from scheduler import ExpressionFissioner, ZeroRemover, SSALoopMerger
 from rewriter import ExpressionRewriter
+from cse import CSEUnpicker
 from coffee.visitors import FindInstances, ProjectExpansion
 
 
@@ -98,6 +99,10 @@ class LoopOptimizer(object):
         elif mode == 'auto-aggressive':
             self._dissect('aggressive')
 
+        # Search for factorization opportunities across temporaries in the kernel
+        if mode > 1 and self.exprs:
+            self._unpick_cse()
+
         # Expression rewriting, expressed as a sequence of AST transformation passes
         for stmt, expr_info in self.exprs.items():
             ew = ExpressionRewriter(stmt, expr_info, self.decls, self.header,
@@ -110,7 +115,6 @@ class LoopOptimizer(object):
                     ew.licm()
 
             elif expr_info.mode == 2:
-                ew.unpickCSE()
                 if expr_info.dimension > 0:
                     ew.replacediv()
                     ew.SGrewrite()
@@ -277,6 +281,14 @@ class LoopOptimizer(object):
         insert_at_elem(self.header.children, self.loop, outer_block)
         # ... scalar-expanding the precomputed symbols
         ast_update_rank(self.loop, precomputed_syms)
+
+    def _unpick_cse(self):
+        """Search for factorization opportunities across temporaries created by
+        common sub-expression elimination. If a gain in operation count is detected,
+        unpick CSE and apply factorization + code motion."""
+        cse_unpicker = CSEUnpicker(self.exprs, self.header, self.hoisted,
+                                   self.decls, self.expr_graph)
+        cse_unpicker.unpick()
 
     def _min_temporaries(self):
         """Remove unnecessary temporaries, thus relieving memory pressure.

--- a/coffee/rewriter.py
+++ b/coffee/rewriter.py
@@ -41,7 +41,6 @@ from coffee.visitors import *
 from hoister import Hoister
 from expander import Expander
 from factorizer import Factorizer
-from cse import CSEUnpicker
 
 
 class ExpressionRewriter(object):
@@ -506,14 +505,6 @@ class ExpressionRewriter(object):
             self.licm()
 
         return self
-
-    def unpickCSE(self):
-        """Search for factorization opportunities across temporaries created by
-        common sub-expression elimination. If a gain in operation count is detected,
-        unpick CSE and apply factorization + code motion."""
-        cse_unpicker = CSEUnpicker(self.stmt, self.expr_info, self.header, self.hoisted,
-                                   self.decls, self.expr_graph)
-        cse_unpicker.unpick()
 
     @staticmethod
     def reset():

--- a/coffee/scheduler.py
+++ b/coffee/scheduler.py
@@ -529,7 +529,7 @@ class ZeroRemover(LoopScheduler):
             def_itspace = [tuple((l.dim, (l.size, 0)) for l, p in nest)]
             nz_bounds = zip(*nz_syms.get(node.symbol, []))
             for i, (r, o, nz_bs) in enumerate(zip(node.rank, node.offset, nz_bounds)):
-                if o[0] != 1 or isinstance(o[1], str) or isinstance(r, int):
+                if o[0] != 1 or isinstance(o[1], str) or is_const_dim(r):
                     # Cannot handle jumps, non-integer offsets, or constant accesses
                     continue
                 try:
@@ -624,7 +624,7 @@ class ZeroRemover(LoopScheduler):
             if not nest:
                 return
 
-            if candidates and nest[-1][0] not in candidates:
+            if nest[-1][0] not in candidates:
                 return
 
             # Track the propagation of non zero-valued blocks: ...
@@ -632,7 +632,7 @@ class ZeroRemover(LoopScheduler):
             itspaces = self._track_nz_expr(expr, nz_syms, nest)
             for i in itspaces:
                 # ... and then through the lvalue (merging overlaps)
-                nz_expr = tuple(dict(i).get(r) for r in sym.rank if not isinstance(r, int))
+                nz_expr = tuple(dict(i).get(r) for r in sym.rank if not is_const_dim(r))
                 if any(j is None for j in nz_expr):
                     break
                 nz_node = list(nz_syms.setdefault(sym.symbol, [nz_expr]))
@@ -669,7 +669,7 @@ class ZeroRemover(LoopScheduler):
         elif isinstance(node, (If, Switch, FunCall)):
             raise RuntimeError("zero blocks tracking: illegal control flow")
 
-    def _reschedule_itspace(self, root):
+    def _reschedule_itspace(self, root, candidates):
         """Consider two statements A and B, and their iteration space. If the
         two iteration spaces have
 
@@ -699,7 +699,7 @@ class ZeroRemover(LoopScheduler):
         """
         nz_info = OrderedDict()
 
-        # 1) Elaborate the initial sparsity pattern of the symbols in /root/
+        # Elaborate the initial sparsity pattern of the symbols in /root/
         nz_syms = defaultdict(list)
         for s, d in self.decls.items():
             if not d.nonzero:
@@ -709,17 +709,11 @@ class ZeroRemover(LoopScheduler):
                 if not np.all(d.init.values[np.ix_(*entries)] == 0.0):
                     nz_syms[s].append(nz_b)
 
-        # 2) Determine the loops that should be analyzed
-        linear_expr_loops = [(l for l in ei.linear_loops) for ei in self.exprs.values()]
-        linear_expr_loops = set(flatten(linear_expr_loops))
-        candidates = [l for l in inner_loops(root)
-                      if not l.is_linear or l in linear_expr_loops]
-
-        # 3) Track the propagation of non zero-valued blocks through symbolic
+        # Track the propagation of non zero-valued blocks through symbolic
         # execution. This populates /nz_info/ and updates /nz_syms/
         self._track_nz_blocks(root, nz_syms, nz_info, candidates=candidates)
 
-        # 4) At this point we know where non-zero blocks are located, so we have
+        # At this point we know where non-zero blocks are located, so we have
         # to create proper loop nests to access them
         new_exprs, new_nz_info = OrderedDict(), OrderedDict()
         for nest, stmt_itspaces in nz_info.items():
@@ -830,24 +824,37 @@ class ZeroRemover(LoopScheduler):
         starting from ``root``. Control flow, in the form of If, Switch, etc.,
         is forbidden."""
 
+        # Avoid rescheduling if zero-valued blocks are too small
         zero_decls = [d for d in self.decls.values() if d.nonzero]
         if self._should_skip(zero_decls):
-            return
+            return {}
 
-        # First, split the main expressions to maximize the impact of the transformation.
-        # This helps if different summands have zero-valued blocks at different offsets
-        elf = ExpressionFissioner(cut=1, loops='none')
-        for stmt, expr_info in self.exprs.items():
-            if expr_info.is_scalar:
-                continue
-            self.exprs.pop(stmt)
-            self.exprs.update(elf.fission(stmt, expr_info))
+        # Determine the analyzable loops (inner loops in which statements have no
+        # read-after-write dependencies)
+        linear_expr_loops = [(l for l in ei.linear_loops) for ei in self.exprs.values()]
+        linear_expr_loops = set(flatten(linear_expr_loops))
+        candidates = [l for l in inner_loops(root) if not l.is_linear or l in linear_expr_loops]
+        candidates = [l for l in candidates if not ExpressionGraph(l.body).has_dependency()]
+        if not candidates:
+            return {}
 
-        # Perform symbolic execution, track the propagation of non zero-valued
-        # blocks, restructure the iteration spaces to avoid useless arithmetic ops
-        nz_syms, nz_info = self._reschedule_itspace(root)
+        if linear_expr_loops & set(candidates):
+            # Split the main expressions to maximize the impact of the rescheduling (this
+            # helps if different summands have zero-valued blocks at different offsets)
+            elf = ExpressionFissioner(cut=1, loops='none')
+            for stmt, expr_info in self.exprs.items():
+                if expr_info.is_scalar:
+                    continue
+                self.exprs.pop(stmt)
+                self.exprs.update(elf.fission(stmt, expr_info))
 
-        # Finally, "inline" the expressions that were originally split
-        self._recombine(nz_info)
+            # Apply the rescheduling
+            nz_syms, nz_info = self._reschedule_itspace(root, candidates)
+
+            # Finally, "inline" the expressions that were originally split, if possible
+            self._recombine(nz_info)
+        else:
+            # Apply the rescheduling
+            nz_syms, nz_info = self._reschedule_itspace(root, candidates)
 
         return nz_syms

--- a/coffee/utils.py
+++ b/coffee/utils.py
@@ -797,7 +797,7 @@ def uniquify(exprs):
     """Iterate over ``exprs`` and return a list of expressions in which duplicates
     have been discarded. This function considers two expressions identical if they
     have the same string representation."""
-    return dict([(str(e), e) for e in exprs]).values()
+    return OrderedDict([(e.urepr, e) for e in exprs]).values()
 
 
 def postprocess(node):

--- a/coffee/utils.py
+++ b/coffee/utils.py
@@ -708,11 +708,9 @@ class ExpressionGraph(object):
         writes = FindInstances(Writer).visit(node, ret=FindInstances.default_retval())
         for type, nodes in writes.items():
             for n in nodes:
-                lvalue = n.lvalue
-                rvalue = n.rvalue
-                if isinstance(rvalue, EmptyStatement):
+                if isinstance(n.rvalue, EmptyStatement):
                     continue
-                self.add_dependency(lvalue, rvalue)
+                self.add_dependency(n.lvalue, n.rvalue)
 
     def add_dependency(self, sym, expr):
         """Add dependency between ``sym`` and symbols appearing in ``expr``."""
@@ -720,6 +718,12 @@ class ExpressionGraph(object):
         expr_symbols = FindInstances(Symbol).visit(expr, ret=retval)[Symbol]
         for es in expr_symbols:
             self.deps.add_edge(sym.symbol, es.symbol)
+
+    def has_dependency(self):
+        """Return True if a read-after-write (raw) or write-after-read (war)
+        dependency appears in the graph, False otherwise."""
+        sources, targets = zip(*self.deps.edges())
+        return True if set(sources) & set(targets) else False
 
     def is_read(self, expr, target_sym=None):
         """Return True if any symbols in ``expr`` is read by ``target_sym``,

